File: sdks/go/retention_test.go
package veritio

import (
	"crypto/ed25519"
	"encoding/hex"
	"reflect"
	"testing"
)

func TestRetentionCheckpointFixtures(t *testing.T) {
	data := loadFixture(t, "retention-checkpoints.json")
	cases := retentionArrayValue(t, data["cases"])
	unsigned, err := CreateRetentionCheckpoint(decodeValue[RetentionCheckpointInput](t, mapValue(t, cases[0])["input"]), nil)
	if err != nil {
		t.Fatalf("CreateRetentionCheckpoint returned error: %v", err)
	}
	if !reflect.DeepEqual(toJSONMap(t, unsigned), mapValue(t, mapValue(t, cases[0])["expected"])) {
		t.Fatalf("unsigned checkpoint did not match fixture")
	}

	signedCase := mapValue(t, cases[1])
	signature := decodeValue[RetentionSignature](t, signedCase["signature"])
	signed, err := CreateRetentionCheckpoint(decodeValue[RetentionCheckpointInput](t, signedCase["input"]), &signature)
	if err != nil {
		t.Fatalf("CreateRetentionCheckpoint returned error: %v", err)
	}
	if signed.Hash != stringValue(t, signedCase["expectedHash"]) {
		t.Fatalf("signed checkpoint hash mismatch")
	}
	key, _ := hex.DecodeString(stringValue(t, data["publicKeyHex"]))
	result := VerifyRetentionCheckpoint(signed, &RetentionVerificationOptions{
		TrustedPublicKey: key,
		RequireSignature: true,
		SignatureVerifier: func(publicKey, signature, message []byte) bool {
			return ed25519.Verify(ed25519.PublicKey(publicKey), message, signature)
		},
	})
	if !result.OK || result.Signature != "valid" {
		t.Fatalf("expected valid signed checkpoint, got %#v", result)
	}
}

func TestRetentionRejectsConstructorVectors(t *testing.T) {
	data := loadFixture(t, "retention-checkpoints.json")
	base := mapValue(t, mapValue(t, retentionArrayValue(t, data["cases"])[0])["input"])
	for _, raw := range retentionArrayValue(t, data["constructorRejections"]) {
		rejection := mapValue(t, raw)
		candidate := retentionCloneMap(base)
		candidate[stringValue(t, rejection["field"])] = rejection["value"]
		if _, err := CreateRetentionCheckpoint(decodeValue[RetentionCheckpointInput](t, candidate), nil); err == nil {
			t.Fatalf("expected rejection for %s", stringValue(t, rejection["name"]))
		}
	}
}

func TestRetentionChainAndAnchoredTail(t *testing.T) {
	data := loadFixture(t, "retention-checkpoints.json")
	cases := retentionArrayValue(t, data["cases"])
	first, _ := CreateRetentionCheckpoint(decodeValue[RetentionCheckpointInput](t, mapValue(t, cases[0])["input"]), nil)
	secondInput := decodeValue[RetentionCheckpointInput](t, mapValue(t, cases[1])["input"])
	secondInput.SignaturePublicKeyFingerprint = ""
	secondInput.FromSequence = 4
	secondInput.ThroughSequence = 4
	gap, _ := CreateRetentionCheckpoint(secondInput, nil)
	result := VerifyRetentionCheckpointChain([]RetentionCheckpoint{first, gap}, nil)
	if result.OK || result.Index != 1 || result.Reason != "range_mismatch" {
		t.Fatalf("expected range mismatch, got %#v", result)
	}
	records := decodeValue[[]AuditRecord](t, data["retainedRecords"])
	if result := VerifyAuditRecordsFromCheckpoint(first, records, nil); !result.OK {
		t.Fatalf("expected anchored tail to verify, got %#v", result)
	}
}

func TestRetentionSignatureFingerprintMismatch(t *testing.T) {
	data := loadFixture(t, "retention-checkpoints.json")
	entry := mapValue(t, retentionArrayValue(t, data["cases"])[1])
	signature := decodeValue[RetentionSignature](t, entry["signature"])
	signature.PublicKeyFingerprint = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
	checkpoint, _ := CreateRetentionCheckpoint(decodeValue[RetentionCheckpointInput](t, entry["input"]), &signature)
	if result := VerifyRetentionCheckpoint(checkpoint, nil); result.Reason != "signature_fingerprint_mismatch" {
		t.Fatalf("expected fingerprint mismatch, got %#v", result)
	}
}

func TestRetentionAllSignedChainReportsValid(t *testing.T) {
	data := loadFixture(t, "retention-checkpoints.json")
	cases := retentionArrayValue(t, data["cases"])
	input := decodeValue[RetentionCheckpointInput](t, mapValue(t, cases[0])["input"])
	input.SignaturePublicKeyFingerprint = mapValue(t, cases[1])["input"].(map[string]any)["signaturePublicKeyFingerprint"].(string)
	signature := decodeValue[RetentionSignature](t, mapValue(t, cases[1])["signature"])
	checkpoint, err := CreateRetentionCheckpoint(input, &signature)
	if err != nil {
		t.Fatalf("CreateRetentionCheckpoint returned error: %v", err)
	}
	key, _ := hex.DecodeString(stringValue(t, data["publicKeyHex"]))
	result := VerifyRetentionCheckpointChain([]RetentionCheckpoint{checkpoint}, &RetentionVerificationOptions{
		TrustedPublicKey: key, RequireSignature: true,
		SignatureVerifier: func(_, _, _ []byte) bool { return true },
	})
	if !result.OK || result.Signature != "valid" {
		t.Fatalf("expected all-signed chain status valid, got %#v", result)
	}
}

func TestRetentionDispositionFixtureAndMismatch(t *testing.T) {
	checkpoints := loadFixture(t, "retention-checkpoints.json")
	dispositions := loadFixture(t, "retention-dispositions.json")
	checkpoint, _ := CreateRetentionCheckpoint(decodeValue[RetentionCheckpointInput](t, mapValue(t, retentionArrayValue(t, checkpoints["cases"])[0])["input"]), nil)
	entry := mapValue(t, retentionArrayValue(t, dispositions["cases"])[0])
	signature := decodeValue[RetentionSignature](t, entry["signature"])
	disposition, err := CreateRetentionDisposition(decodeValue[RetentionDispositionInput](t, entry["input"]), &signature)
	if err != nil || disposition.Hash != stringValue(t, entry["expectedHash"]) {
		t.Fatalf("disposition fixture mismatch: %v", err)
	}
	for _, raw := range retentionArrayValue(t, dispositions["mismatchRejections"]) {
		rejection := mapValue(t, raw)
		input := decodeValue[RetentionDispositionInput](t, entry["input"])
		input.SignaturePublicKeyFingerprint = ""
		switch stringValue(t, rejection["field"]) {
		case "checkpointHash":
			input.CheckpointHash = stringValue(t, rejection["value"])
		case "archiveRootHash":
			input.ArchiveRootHash = stringValue(t, rejection["value"])
		}
		mismatch, _ := CreateRetentionDisposition(input, nil)
		if result := VerifyRetentionDisposition(mismatch, checkpoint, nil); result.Reason != "checkpoint_mismatch" {
			t.Fatalf("expected checkpoint mismatch, got %#v", result)
		}
	}
}

func retentionCloneMap(input map[string]any) map[string]any {
	output := make(map[string]any, len(input))
	for key, value := range input {
		output[key] = value
	}
	return output
}

func retentionArrayValue(t *testing.T, value any) []any {
	t.Helper()
	typed, ok := value.([]any)
	if !ok {
		t.Fatalf("expected array, got %#v", value)
	}
	return typed
}

File: sdks/typescript/src/__tests__/retention.test.ts
import { describe, expect, test } from "bun:test";
import { createPublicKey, verify } from "node:crypto";
import { join } from "node:path";
import type { AuditRecord } from "../index";
import {
  createRetentionCheckpoint,
  createRetentionDisposition,
  hashRetentionCheckpoint,
  hashRetentionDisposition,
  type RetentionCheckpoint,
  verifyAuditRecordsFromCheckpoint,
  verifyRetentionCheckpoint,
  verifyRetentionCheckpointChain,
  verifyRetentionDisposition,
} from "../retention";

const CONFORMANCE_DIR = join(import.meta.dir, "../../../../spec/conformance");

async function fixture<T>(name: string): Promise<T> {
  return (await Bun.file(join(CONFORMANCE_DIR, name)).json()) as T;
}

function hexBytes(value: string): Uint8Array {
  return Uint8Array.from(Buffer.from(value, "hex"));
}

function ed25519Verifier(publicKey: Uint8Array, signature: Uint8Array, message: Uint8Array): boolean {
  const spkiPrefix = Buffer.from("302a300506032b6570032100", "hex");
  const key = createPublicKey({ key: Buffer.concat([spkiPrefix, publicKey]), format: "der", type: "spki" });
  return verify(null, message, key, signature);
}

describe("retention checkpoint protocol", () => {
  test("matches literal checkpoint hashes and detached signature fixture", async () => {
    const data = await fixture<any>("retention-checkpoints.json");
    const unsigned = createRetentionCheckpoint(data.cases[0].input);
    const signed = createRetentionCheckpoint(data.cases[1].input, data.cases[1].signature);

    expect(unsigned).toEqual(data.cases[0].expected);
    expect(hashRetentionCheckpoint(unsigned)).toBe(data.cases[0].expected.hash);
    expect(signed.hash).toBe(data.cases[1].expectedHash);
    expect(
      verifyRetentionCheckpoint(signed, {
        trustedPublicKey: hexBytes(data.publicKeyHex),
        signatureVerifier: ed25519Verifier,
        requireSignature: true,
      }),
    ).toEqual({ ok: true, signature: "valid" });
  });

  test("rejects invalid ids, unsafe integers, and non-millisecond timestamps", async () => {
    const data = await fixture<any>("retention-checkpoints.json");
    for (const rejection of data.constructorRejections) {
      expect(() =>
        createRetentionCheckpoint({ ...data.cases[0].input, [rejection.field]: rejection.value }),
      ).toThrow();
    }
  });

  test("requires exact epoch and range continuity", async () => {
    const data = await fixture<any>("retention-checkpoints.json");
    const first = createRetentionCheckpoint(data.cases[0].input);
    const { signaturePublicKeyFingerprint: _fingerprint, ...secondInput } = data.cases[1].input;
    const gap = createRetentionCheckpoint({ ...secondInput, fromSequence: 4, throughSequence: 4 });

    expect(verifyRetentionCheckpointChain([first, gap])).toEqual({
      ok: false,
      index: 1,
      reason: "range_mismatch",
      signature: "absent",
    });
  });

  test("reports valid when every checkpoint signature verifies", async () => {
    const data = await fixture<any>("retention-checkpoints.json");
    const fingerprint = data.cases[1].input.signaturePublicKeyFingerprint;
    const checkpoint = createRetentionCheckpoint(
      { ...data.cases[0].input, signaturePublicKeyFingerprint: fingerprint },
      data.cases[1].signature,
    );

    expect(
      verifyRetentionCheckpointChain([checkpoint], {
        trustedPublicKey: hexBytes(data.publicKeyHex),
        signatureVerifier: () => true,
        requireSignature: true,
      }),
    ).toEqual({ ok: true, signature: "valid" });
  });

  test("rejects a signature fingerprint not bound into the checkpoint hash", async () => {
    const data = await fixture<any>("retention-checkpoints.json");
    const signed = createRetentionCheckpoint(data.cases[1].input, {
      ...data.cases[1].signature,
      publicKeyFingerprint: "f".repeat(64),
    });

    expect(verifyRetentionCheckpoint(signed).reason).toBe("signature_fingerprint_mismatch");
  });

  test("verifies a retained audit tail from the checkpoint anchor", async () => {
    const data = await fixture<any>("retention-checkpoints.json");
    const checkpoint = createRetentionCheckpoint(data.cases[0].input);

    expect(verifyAuditRecordsFromCheckpoint(checkpoint, data.retainedRecords as AuditRecord[])).toEqual({
      ok: true,
      signature: "absent",
    });
    expect(
      verifyAuditRecordsFromCheckpoint(checkpoint, [
        { ...data.retainedRecords[0], previousHash: "f".repeat(64) },
      ] as AuditRecord[]).reason,
    ).toBe("previous_hash_mismatch");
  });
});

describe("retention disposition protocol", () => {
  test("matches the literal hash and binds the referenced checkpoint", async () => {
    const checkpoints = await fixture<any>("retention-checkpoints.json");
    const data = await fixture<any>("retention-dispositions.json");
    const checkpoint = createRetentionCheckpoint(checkpoints.cases[0].input);
    const disposition = createRetentionDisposition(data.cases[0].input, data.cases[0].signature);

    expect(disposition.hash).toBe(data.cases[0].expectedHash);
    expect(hashRetentionDisposition(disposition)).toBe(data.cases[0].expectedHash);
    expect(
      verifyRetentionDisposition(disposition, checkpoint, {
        trustedPublicKey: hexBytes(checkpoints.publicKeyHex),
        signatureVerifier: ed25519Verifier,
        requireSignature: true,
      }),
    ).toEqual({ ok: true, signature: "valid" });
  });

  test("fails closed on disposition/checkpoint mismatches", async () => {
    const checkpoints = await fixture<any>("retention-checkpoints.json");
    const data = await fixture<any>("retention-dispositions.json");
    const checkpoint = createRetentionCheckpoint(checkpoints.cases[0].input);
    const { signaturePublicKeyFingerprint: _fingerprint, ...baseInput } = data.cases[0].input;

    for (const rejection of data.mismatchRejections) {
      const disposition = createRetentionDisposition({ ...baseInput, [rejection.field]: rejection.value });
      expect(verifyRetentionDisposition(disposition, checkpoint).reason).toBe("checkpoint_mismatch");
    }
  });
});
